pointgrey_camera_driver/src/PointGreyCamera.cpp

```diff
@@ -47,1107 +52,845 @@ PointGreyCamera::~PointGreyCamera()
+
+  return result;
 }
 
-void PointGreyCamera::start()
+int PointGreyCamera::start()
 {
-  if(cam_.IsConnected() && !captureRunning_)
+  int result = 0;
+
+  try
+  {
+    // Check if camera is connected
+    if(pCam_ && !captureRunning_)
+    {
+      // Start capturing images
+      pCam_->BeginAcquisition();
+      captureRunning_ = true;
+    }
+  }
+  catch (Spinnaker::Exception &e)
   {
-    // Start capturing images
-    Error error = cam_.StartCapture();
-    PointGreyCamera::handleError("PointGreyCamera::start Failed to start capture", error);
-    captureRunning_ = true;
+    ROS_ERROR_STREAM("PointGreyCamera::start Failed to start capture with Error: " << e.what());
+    int result = -1;
   }
+  return result;
 }
 
+
 bool PointGreyCamera::stop()
 {
-  if(cam_.IsConnected() && captureRunning_)
+  if (pCam_ && captureRunning_)
   {
     // Stop capturing images
-    captureRunning_ = false;
-    Error error = cam_.StopCapture();
-    PointGreyCamera::handleError("PointGreyCamera::stop Failed to stop capture", error);
-    return true;
+    try
+    {
+      captureRunning_ = false;
+      pCam_->EndAcquisition();
+      return true;
+    }
+    catch (Spinnaker::Exception &e)
+    {
+      ROS_ERROR_STREAM("PointGreyCamera::stop Failed to stop capture with Error: " << e.what());
+      return false;
+    }
   }
   return false;
 }
 
-void PointGreyCamera::grabImage(sensor_msgs::Image &image, const std::string &frame_id)
+
+int PointGreyCamera::grabImage(sensor_msgs::Image &image, const std::string &frame_id)
 {
+  int result = 0;
+
   boost::mutex::scoped_lock scopedLock(mutex_);
-  if(cam_.IsConnected() && captureRunning_)
+
+  // Check if Camera is connected and Running
+  if(pCam_ && captureRunning_)
   {
-    // Make a FlyCapture2::Image to hold the buffer returned by the camera.
-    Image rawImage;
-    // Retrieve an image
-    Error error = cam_.RetrieveBuffer(&rawImage);
-    PointGreyCamera::handleError("PointGreyCamera::grabImage Failed to retrieve buffer", error);
-    metadata_ = rawImage.GetMetadata();
-
-    // Set header timestamp as embedded for now
-    TimeStamp embeddedTime = rawImage.GetTimeStamp();
-    image.header.stamp.sec = embeddedTime.seconds;
-    image.header.stamp.nsec = 1000 * embeddedTime.microSeconds;
-
-    // Check the bits per pixel.
-    uint8_t bitsPerPixel = rawImage.GetBitsPerPixel();
-
-    // Set the image encoding
-    std::string imageEncoding = sensor_msgs::image_encodings::MONO8;
-    BayerTileFormat bayer_format = rawImage.GetBayerTileFormat();
-    if(isColor_ && bayer_format != NONE)
+    // Handle "Image Retrieval" Exception
+    try
     {
-      if(bitsPerPixel == 16)
+      Spinnaker::ImagePtr image_ptr = pCam_->GetNextImage();
+      std::string format(image_ptr->GetPixelFormatName());
+
+      // Set Image Time Stamp
+      image.header.stamp.sec = image_ptr->GetTimeStamp() * 1e-9;
+      image.header.stamp.nsec = image_ptr->GetTimeStamp();
+
+      if (image_ptr->IsIncomplete())
```

Comment:
The point is to provide notification that the image is incomplete. This could be added to diagnostics. The image isn't used if incomplete so no need to discard anyway.